exp/addons/internal/controllers/clusterresourceset_controller_test.go

```diff
@@ -588,4 +598,389 @@ metadata:
+	})
+
+	t.Run("Should reconcile a ClusterResourceSet with ApplyOnce strategy even when one of the resources already exist", func(t *testing.T) {
+		g := NewWithT(t)
+		ns := setup(t, g)
+		defer teardown(t, g, ns)
+
+		t.Log("Updating the cluster with labels")
+		testCluster.SetLabels(labels)
+		g.Expect(env.Update(ctx, testCluster)).To(Succeed())
+
+		t.Log("Creating resource CM before creating CRS")
+		// This CM is defined in the data of "configmapName", which is included in the
+		// CRS we will create in this test
+		resourceConfigMap1 := configMap(
+			resourceConfigMap1Name,
+			resourceConfigMapsNamespace,
+			map[string]string{
+				"created": "before CRS",
+			},
+		)
+		g.Expect(env.Create(ctx, resourceConfigMap1)).To(Succeed())
+
+		t.Log("Creating a ClusterResourceSet instance that has same labels as selector")
+		clusterResourceSet := &addonsv1.ClusterResourceSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      clusterResourceSetName,
+				Namespace: ns.Name,
+			},
+			Spec: addonsv1.ClusterResourceSetSpec{
+				Strategy: string(addonsv1.ClusterResourceSetStrategyApplyOnce),
+				ClusterSelector: metav1.LabelSelector{
+					MatchLabels: labels,
+				},
+				Resources: []addonsv1.ResourceRef{{Name: configmapName, Kind: "ConfigMap"}, {Name: secretName, Kind: "Secret"}},
+			},
+		}
+
+		g.Expect(env.Create(ctx, clusterResourceSet)).To(Succeed())
+
+		t.Log("Checking resource ConfigMap 1 hasn't been updated")
+		resourceConfigMap1Key := client.ObjectKey{
+			Namespace: resourceConfigMapsNamespace,
+			Name:      resourceConfigMap1Name,
+		}
+		g.Eventually(configMapHasBeenUpdated(env, resourceConfigMap1Key, resourceConfigMap1), timeout).Should(Succeed())
+
+		t.Log("Verifying resource ConfigMap 2 has been created")
+		resourceConfigMap2Key := client.ObjectKey{
+			Namespace: resourceConfigMapsNamespace,
+			Name:      resourceConfigMap2Name,
+		}
+		g.Eventually(func() error {
+			cm := &corev1.ConfigMap{}
+			return env.Get(ctx, resourceConfigMap2Key, cm)
+		}, timeout).Should(Succeed())
+	})
+
+	t.Run("Should reconcile a ClusterResourceSet with ApplyOnce strategy even when there is an error, after the error has been corrected", func(t *testing.T) {
+		// To trigger an error in the middle of the reconciliation, we'll define a an object in a namespace that doesn't yet exist.
+		// We'll expect the CRS to reconcile all other objects except that one and bubble up the error.
+		// Once that happens, we'll go ahead a create the namespace. Then we'll expect the CRS to, eventually, create that remaining object.
+
+		g := NewWithT(t)
+		ns := setup(t, g)
+		defer teardown(t, g, ns)
+
+		t.Log("Updating the cluster with labels")
+		testCluster.SetLabels(labels)
+		g.Expect(env.Update(ctx, testCluster)).To(Succeed())
+
+		t.Log("Updating the test config map with the missing namespace resource")
+		missingNamespace := randomNamespaceForTest(t)
```

Comment:
As the linter is complaining about the way you're generating the name anyway. WDYT about just using something like this?

```suggestion
		missingNamespace := fmt.Sprintf("clusterresourceset-apply-once-%s", util.RandomString(6))
```

Reply:
I can change it if you prefer but I thought it was a useful function to have.

btw fixed the linter

Reply:
Fine for me to keep just seems more complicated to have our own random func here vs. using the util that we use in a lot of other places

Up to you